### PR TITLE
daemon: ucred based peer enrichment (p3)

### PR DESCRIPTION
**Pull request by ernestl**
**Conflicts with  https://github.com/canonical/snapd/pull/17608 - can only complete after that lands.**

Build `seclog.Peer` from Unix socket credentials plus best-effort `/proc` enrichment for AUTHZ events.

Peer fields: 
- `socket`, `uid`, `pid` from ucred; 
- `exe` from `/proc/<pid>/exe`; `snap` and `runnable` from a security tag. The tag is AppArmor when that label is present and not `unconfined`, otherwise the cgroup security tag. LSM labels and the cgroup tag are not stored on Peer.

[Design decisions](https://docs.google.com/document/d/1fl7Gy3GJVu8VwUtT1_LWwZ7DEAW9sUMAzF3XGrcZ_d4/edit?tab=t.u6gbi3cr8z7t#bookmark=id.8o0k0akbm3br)

`seclogPeerFromUcred` is unused on the daemon request path in this PR. That is expected for p3; AUTHZ emission in p4 is the intended caller.

`osutil.ReadProcLSMCurrent` is LSM-specific and lives in osutil so `sandbox/apparmor` and `sandbox/selinux` can share the stacked-LSM `/proc` read. `package sandbox` already imports apparmor, so those packages cannot share a helper through sandbox or `sandbox/lsm` without an import cycle.

Jira: https://warthogs.atlassian.net/browse/SNAPDENG-37105
Spec: [SD236- Security Event Logging](https://docs.google.com/document/d/1fl7Gy3GJVu8VwUtT1_LWwZ7DEAW9sUMAzF3XGrcZ_d4/edit?tab=t.0#bookmark=id.n4uji26iddcb)

### CODE REVIEW

File: daemon/seclog.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/sandbox/cgroup"
+	"github.com/snapcore/snapd/sandbox/lsm"
+	"github.com/snapcore/snapd/seclog"
+	"github.com/snapcore/snapd/snap/naming"
+)
+
+var (
+	securityLabelsFromPid = lsm.SecurityLabelsFromPid
```
Comment (ernestl) [inline daemon/seclog.go]: Correct naming

File: daemon/seclog.go
Comment (ernestl) [inline daemon/seclog.go:1]: This exists to prevent circular imports due to having using daemon types in seclog package.

File: sandbox/apparmor/process.go [hunk 2]
```diff
@@ -21,30 +21,29 @@ package apparmor
...
+	}
+
+	label, err := osutil.ReadProcLSMCurrent(pid, "apparmor")
+	if err != nil {
 		return "", err
 	}
-	label := strings.TrimRight(string(contents), "\n")
-	// Trim off the mode
+	if label == "" {
+		return "unconfined", nil
```
Comment (zyga) [inline sandbox/apparmor/process.go]: You can have a label that is not apparmor label that is non empty that is unconfined. As long as we only call this on apparmor systems, it should be okay though

File: snap/info.go [hunk 3]
```diff
@@ -1363,6 +1363,7 @@ type AppInfo struct {
 }
 
 // Runnable returns a Runnable for this app.
+// CommandName matches [naming.SecurityTag.CommandName].
```
Comment (bboozzoo) [inline snap/info.go]: this isn't enforceable without a unit test. Maybe we should actually switch to using `snap/naming` rather than keep deriving both values in different packages and hope they won't go out of sync.

File: daemon/seclog.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+// Component hooks omit the snap name, which is recorded on [seclog.Peer.Snap].
+func seclogRunnable(tag naming.SecurityTag) string {
+	if _, ok := tag.(naming.AppSecurityTag); ok {
+		return "app." + tag.CommandName()
+	}
+	if hook, ok := tag.(naming.HookSecurityTag); ok && hook.ComponentName() != "" {
+		return hook.ComponentName() + ".hook." + hook.HookName()
+	}
+	return tag.CommandName()
+}
```
Comment (bboozzoo) [inline daemon/seclog.go]: isn't the naming pattern specific to seclog.Peer and we're effectively leaking some implementation detail into `daemon`? Perhaps you could add:

```
// in seclog package

type SnapRunnable struct {
    // InstanceName of the snap
    InstanceName string
    // AppCommandName is set if runnable describes an app
    AppCommandName string
    // HookName is set if runnable describes a hook
    HookName string
    // HookComponentName if runnable describes a component hook
    HookComponentName string
}

func (p *Peer) WithRunnable(r SnapRunnable) *Peer {
    p.Snap = r.InstanceName
    if r.HookName == "" {
        seclogRunnable = "app." + r.CommandName
    } else {
        if r.ComponentName == "" {
            seclogRunnable = fmt.Sprintf("hook.%s", r.Hookname)
        } else {
            seclogRunnable = fmt.Sprintf("%s+%s.hook.%s", p.Snap, r.ComponentName)
        }
    }
    p.Runnable = seclogRunnable
    return p
}
```

and then since you no longer need `tag.CommandName()`, you can avoid the issue with naming entirely.

File: snap/info.go (hunk 3 above)
Comment (bboozzoo) [reply snap/info.go]: Though I think this can be avoided if you pass around a helper struct describing the runnable and let `seclog.Peer` be filled using a helper in the `seclog` package.

File: daemon/seclog.go (hunk 4 above)
Comment (ernestl) [reply daemon/seclog.go]: Thanks. That format belongs to the peer log, so it should not live in the daemon.

I skipped WithRunnable. A security tag always has both a snap and a runnable, and a helper struct would only carry AppName, HookName, and ComponentName, which the tag already has.

seclog.RunnableFromSecurityTag takes the tag and builds the log string from those methods: app.<name>, hook.<name>, and <component>.hook.<name>. The snap instance stays on Peer.Snap.

That also removes CommandName, so there is no second copy of the snap.Runnable command string to keep in sync.

File: snap/info.go (hunk 3 above)
Comment (ernestl) [reply snap/info.go]: See: https://github.com/canonical/snapd/pull/17283#discussion_r4092403825

**Review by bboozzoo — APPROVED** [review]
LGTM

File: daemon/seclog.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+			UID: seclog.PeerNobody,
+			PID: seclog.PeerNoProcess,
+		}
+	}
+	peer := seclog.Peer{
+		Socket: ucred.Socket,
+		UID:    ucred.Uid,
+		PID:    ucred.PIDForPolkit, // SO_PEERCRED captured at accept
+	}
+	if exe, err := ucred.UntrustedProcessExeName(); err == nil {
```
Comment (zyga) [inline daemon/seclog.go]:
```suggestion
    // Note that untrusted is just that we have low confidence in the name, not that the process itself is not trusted explicitly.
	if exe, err := ucred.UntrustedProcessExeName(); err == nil {
```

File: seclog/eventdata.go [hunk 6]
```diff
@@ -92,61 +93,57 @@ type SnapdUser struct {
...
-	// and [PeerSecurityLabelSELinux]. Omit unavailable keys.
-	SecurityLabels map[string]string `json:"security_labels"`
-	// CgroupLabel is the snap cgroup label of the peer process (e.g.
-	// snap.<instance>.<app>). [unknown] when unavailable.
-	CgroupLabel string `json:"cgroup_label"`
-	// Snap is the snap instance name of the peer process, typically derived
-	// from the AppArmor entry in [Peer.SecurityLabels]. [unknown] when unavailable.
+	// Snap is the snap instance name of the peer process, from its snap
+	// security tag. [unknown] when unavailable.
 	Snap string `json:"snap"`
```
Comment (zyga) [inline seclog/eventdata.go]: Snap instance type?

File: daemon/seclog_test.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+	peer := daemon.SeclogPeerFromUcred(ucred)
+
+	c.Check(peer, DeepEquals, seclog.Peer{
+		Socket:   "/run/snapd.socket",
+		UID:      1000,
+		PID:      4242,
+		Snap:     "firefox",
+		Runnable: "app.firefox",
+	})
+}
```
Comment (zyga) [inline daemon/seclog_test.go]: Please add a small test for a component hook as well, and for a snap with a non-empty instance key (and a component) so that we can see all of this checks out.

File: seclog/eventdata_test.go [hunk 8]
```diff
@@ -88,7 +89,35 @@ func (s *SecLogSuite) TestPeerString(c *C) {
...
+		tag:      "snap.firefox.firefox",
+		runnable: "app.firefox",
+	}, {
+		tag:      "snap.firefox_foo.firefox",
+		runnable: "app.firefox",
+	}, {
+		tag:      "snap.mysnap.hook.install",
+		runnable: "hook.install",
+	}, {
+		tag:      "snap.mysnap+comp.hook.install",
```
Comment (zyga) [inline seclog/eventdata_test.go]: Let's use something other than `comp` as it's a component hook or a hook of component named `comp`.

File: seclog/eventdata_test.go [hunk 9]
```diff
@@ -88,7 +89,35 @@ func (s *SecLogSuite) TestPeerString(c *C) {
...
+		runnable: "comp.hook.install",
```
Comment (zyga) [inline seclog/eventdata_test.go]: We need to indicate _which_ component it is.

Do we want this to contain the component name or not?

```
// $var and everything else is literal.
comp.hook.$hook <- we don't know which $comp
$comp.hook.install
comp.$comp.hook.$hook <- maybe this?
comp.hook.$comp.$hook
```

**Review by zyga — APPROVED** [review]
+1 but please fix or make sure you are happy with the component name + component hook name + instance key (probably in snap (really instnace) name already).
